File: js/test/typescript/Ice/facets/TestI.ts
// **********************************************************************
//
// Copyright (c) 2003-2018 ZeroC, Inc. All rights reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICE_LICENSE file included in this distribution.
//
// **********************************************************************

import {Ice} from "ice";
import {Test} from "./Test"
import {TestHelper} from "../../../Common/TestHelper"
const test = TestHelper.test;

export class DI extends Test.D
{
    callA(current:Ice.Current):string
    {
        return "A";
    }

    callB(current:Ice.Current):string
    {
        return "B";
    }

    callC(current:Ice.Current):string
    {
        return "C";
    }

    callD(current:Ice.Current):string
    {
        return "D";
    }
}

export class EmptyI extends Test.Empty
{
}

export class FI extends Test.F
{
    callE(current:Ice.Current):string
    {
        return "E";
    }

    callF(current:Ice.Current):string
    {
        return "F";
    }
}

export class HI extends Test.H
{
    callG(current:Ice.Current):string
    {
        return "G";
    }

    callH(current:Ice.Current):string
    {
        return "H";
    }

    shutdown(current:Ice.Current):void
    {
        current.adapter.getCommunicator().shutdown();
    }
}

File: cpp/test/IceBridge/simple/TestI.cpp
//
// Copyright (c) ZeroC, Inc. All rights reserved.
//

#include "TestI.h"
#include "Ice/Ice.h"
#include "TestHelper.h"

using namespace std;
using namespace Test;

namespace
{
    Ice::Identity callbackId = {"callback", ""};
}

void
MyClassI::callCallbackAsync(function<void()> response, function<void(exception_ptr)> error, const Ice::Current& current)
{
    checkConnection(current.con);
    auto prx = current.con->createProxy<CallbackPrx>(callbackId);

    prx->pingAsync(
        [response = std::move(response)]() { response(); },
        [error = std::move(error)](exception_ptr e) { error(e); });
}

void
MyClassI::getCallbackCountAsync(
    function<void(int)> response,
    function<void(exception_ptr)> error,
    const Ice::Current& current)
{
    checkConnection(current.con);
    auto prx = current.con->createProxy<CallbackPrx>(callbackId);

    prx->getCountAsync(
        [response = std::move(response)](int count) { response(count); },
        [error = std::move(error)](exception_ptr e) { error(e); });
}

int
MyClassI::getConnectionCount(const Ice::Current& current)
{
    checkConnection(current.con);
    return static_cast<int>(_connections.size());
}

string
MyClassI::getConnectionInfo(const Ice::Current& current)
{
    checkConnection(current.con);
    return current.con->toString();
}

void
MyClassI::closeConnection(bool forceful, const Ice::Current& current)
{
    checkConnection(current.con);
    if (forceful)
    {
        current.con->abort();
    }
    else
    {
        current.con->close(nullptr, nullptr);
    }
}

void
MyClassI::datagram(const Ice::Current& current)
{
    checkConnection(current.con);
    test(current.con->getEndpoint()->getInfo()->datagram());
    ++_datagramCount;
}

int
MyClassI::getDatagramCount(const Ice::Current& current)
{
    checkConnection(current.con);
    return _datagramCount;
}

void
MyClassI::callDatagramCallback(const Ice::Current& current)
{
    checkConnection(current.con);
    test(current.con->getEndpoint()->getInfo()->datagram());
    current.con->createProxy<CallbackPrx>(callbackId)->datagram();
}

void
MyClassI::getCallbackDatagramCountAsync(
    function<void(int)> response,
    function<void(exception_ptr)> error,
    const Ice::Current& current)
{
    checkConnection(current.con);
    auto prx = current.con->createProxy<CallbackPrx>(callbackId);

    prx->getDatagramCountAsync(
        [response = std::move(response)](int count) { response(count); },
        [error = std::move(error)](auto e) { error(e); });
}

void
MyClassI::shutdown(const Ice::Current& current)
{
    checkConnection(current.con);
    current.adapter->getCommunicator()->shutdown();
}

void
MyClassI::removeConnection(const shared_ptr<Ice::Connection>& con)
{
    lock_guard<mutex> lg(_lock);
    _connections.erase(con);
}

void
MyClassI::checkConnection(const shared_ptr<Ice::Connection>& con)
{
    lock_guard<mutex> lg(_lock);
    if (_connections.find(con) == _connections.end())
    {
        _connections.insert(make_pair(con, 0));
        con->setCloseCallback([self = shared_from_this()](const auto& c) { self->removeConnection(c); });
    }
}
